### ISSUE TITLE
just a question

### ISSUE THREAD

**Issue by happyshows** [gluc/data.tree#7]
Hi,

Thanks for the great work, can you explain a little more on the value of providing traversal function in the package while igraph may do it faster for a larger network? I didn't get it after reading the vignettes.

My painpoint is sometime I need to create list of list objects from data frame. (collapse a column and make it the name of parent list, and collapse another recursively) Is there any convenient way to use this package to do the job? Something like var <- to.data.tree(df)  %>% toList(columnA, columnB, ...)

Again, this is not to discredit your great work, just want to understand it better.

**Comment by gluc** [comment]
Hi,
Thanks for your questions. I have just added support for conversion from data.frame. It's available in the latest from github. e.g.

```{r}
# get a data.frame that we can use to demonstrate as.Node:
acmedf <- as.data.frame(acme, row.names = NULL, optional = FALSE, 'p', 'cost', 'pathString')
# convert it back to a tree
acme2 <- as.Node(acmedf, na.rm = TRUE)
```

Let me know if it suites you.

Regarding your other question: igraph is indeed a fine library. But it's *huge*. It covers entire graph theory, while data.tree only works for regular trees. As a result, I hope that the learning curve for data.tree is much smaller. 
Also, for data.tree, the focus is on a *data type* , rather than on a general library. It's easy to convert to and from other data types, it's fairly easy to subclass, and you can decorate a Node dynamically with attributes and functions. 
In consequence, data.tree doesn't implement any algorithms (assuming we agree that tree traversal is not an algorithm). 

In my main field (finance), trees are very common, and they are often rather small (e.g. an Asset Class classification tree could have 20 nodes and 3 levels). For such applications, flexibility and simplicity are more important than computation speed.

The reason why the package implements traversal functions: 
1. I needed them internally for the conversion methods
2. While it is true that for many applications, the traversal order is of no importance, there are very common applications where this is not the case. Take the expected cost example in the vignette: You can only calculate a Node's expected cost if it's children's expected cost have been calculated previously.
In summary, I would argue that tree traversal is one of the most fundamental things you can do with a tree, and thus data.tree should definitely implement is.

Hope all this makes sense, but pls do let me know if you think it doesn't.

Just out of interest, and as I'm by no means an igraph expert: How would you implement a, say, post-order traversal in igraph?

**Comment by happyshows** [comment]
Thanks for the explanation. I'm no igraph expert either, but I guess answer to your question is graph.bfs / dfs + call back function.

**Comment by gluc** [comment]
Yes, I guess the dfs and the right combination of call backs would work, but e.g. for In-order your callbacks would have to keep track of the already passed nodes in order to know whether they have to do something. Anyway, I was just wondering. Thx for the hint.
